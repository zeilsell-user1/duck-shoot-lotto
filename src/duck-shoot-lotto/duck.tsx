import { useEffect, useState } from "react";

import duck from "./../assets/yellow-duck.png";

interface DuckProps {
  number: number;
  onClick: (number: number) => void;
}

export const Duck = ({ number, onClick }: DuckProps): JSX.Element => {
  const [clicked, setClicked] = useState(false);

  useEffect(() => {
    setClicked(false);
  }, [number]);

  const handleClick = (): void => {
    if (!clicked) {
      onClick(number);
      setClicked(true);
    }
  };

  return (
    <div className={`duck ${clicked ? "disabled" : ""}`} onClick={handleClick}>
      <img src={duck} alt="duck" />
      <div className="number">{number}</div>
    </div>
  );
};
