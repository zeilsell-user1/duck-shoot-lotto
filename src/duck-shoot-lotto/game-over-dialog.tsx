import type { FC } from "react";

interface GameOverDialogProps {
  selectedNumbers: number[];
  onRestart: () => void;
}

export const GameOverDialog: FC<GameOverDialogProps> = ({
  selectedNumbers,
  onRestart,
}) => {
  return (
    <div className="dialog-container">
      <div className="dialog">
        <h2>Game Over</h2>
        <p> You have selected : {selectedNumbers.join(", ")} </p>
        <button onClick={onRestart}>Close</button>
      </div>
    </div>
  );
};
