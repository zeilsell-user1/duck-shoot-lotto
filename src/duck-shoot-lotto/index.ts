export { DuckShootLotto } from "./duck-shoot-lotto";
