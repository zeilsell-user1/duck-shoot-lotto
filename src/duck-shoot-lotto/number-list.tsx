interface NumberListProps {
  numbers: number[];
}

export const NumberList = ({ numbers }: NumberListProps): JSX.Element => {
  return (
    <div>
      <p> You have selected : {numbers.join(", ")} </p>
    </div>
  );
};
