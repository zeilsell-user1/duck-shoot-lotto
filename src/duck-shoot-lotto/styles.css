
.board {
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  grid-gap: 10px;
  margin-bottom: 20px;
}

.game {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.ducks {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  max-width: 800px;
  margin-top: 20px;
}

.duck {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background-color: #fff;
  font-size: 1.5rem;
  font-weight: bold;
  cursor: pointer;
  margin: 5px;
  box-shadow: 0px 2px 6px rgba(0, 0, 0, 0.1);
}

.duck:hover {
  background-color: #f8f8f8;
}

.duck.selected {
  background-color: #00bfff;
  color: #fff;
}

.duck.disabled {
  opacity: 0.5;
  cursor: default;
  pointer-events: none;
}

.buttons {
  display: flex;
  justify-content: center;
  margin-top: 20px;
}

button {
  padding: 10px 20px;
  border: none;
  border-radius: 5px;
  background-color: #fc0;
  color: #333;
  font-size: 16px;
  cursor: pointer;
  margin: 0 10px;
  transition: background-color 0.3s ease;
}

button:hover {
  background-color: #cc0;
}

button:active {
  background-color: #990;
}
