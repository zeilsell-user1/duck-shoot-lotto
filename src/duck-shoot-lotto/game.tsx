import { useState } from "react";
import { Duck } from "./duck";

interface GameProps {
  onFinish: (numbers: number[]) => void;
}

export const Game = ({ onFinish }: GameProps): JSX.Element => {
  const [selectedDucks, setSelectedDucks] = useState<number[]>([]);

  function handleDuckClick(number: number): void {
    if (selectedDucks.includes(number)) {
      setSelectedDucks(selectedDucks.filter((n) => n !== number));
    } else if (selectedDucks.length < 6) {
      setSelectedDucks([...selectedDucks, number]);
    }
  }

  function handleFinishClick(): void {
    if (selectedDucks.length === 6) {
      onFinish(selectedDucks);
    }
  }

  const ducks: JSX.Element[][] = [];
  for (let i = 1; i <= 7; i++) {
    const row: JSX.Element[] = [];
    for (let j = 1; j <= 7; j++) {
      const number = (i - 1) * 7 + j;
      const disabled = selectedDucks.length === 6 && !selectedDucks.includes(number);
      row.push(
        <Duck key={number} number={number} disabled={disabled} onClick={handleDuckClick} />
      );
    }
    ducks.push(row);
  }

  return (
    <div className="game">
      <div className="board">{ducks}</div>
      <button className="finish-button" onClick={handleFinishClick}>
        Finish
      </button>
    </div>
  );
};
