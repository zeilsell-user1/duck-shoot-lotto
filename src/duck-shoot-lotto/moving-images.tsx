import { useEffect,useState } from "react";

import { Duck } from "./duck";

interface MovingImagesProps {
  width:number;
  height:number;
  onMovingImageClick: (number: number) => void;
  availableNumbers: number[];
}

export const MovingImages = ({
  width,
  height,
  onMovingImageClick,
  availableNumbers,
}: MovingImagesProps): JSX.Element => {
  const [duck1Position, setDuck1Position] = useState({ x: 0, y: 0 });
  const [duck2Position, setDuck2Position] = useState({ x: 0, y: 0 });
  const [duck3Position, setDuck3Position] = useState({ x: 0, y: 0 });
  const [duck1Number, setDuck1Number] = useState<number>(1);
  const [duck2Number, setDuck2Number] = useState<number>(2);
  const [duck3Number, setDuck3Number] = useState<number>(3);

  const handleDuckClick = (number: number): void => {
    onMovingImageClick(number);
  };

  const getRandomNumbers = (array: number[], count: number): number[] => {
    const randomNumbers: number[] = [];
    let i = 0;

    while (i < count) {
      const randomIndex = Math.floor(Math.random() * array.length);
      if (!randomNumbers.includes(array[randomIndex])) {
        randomNumbers.push(array[randomIndex]);
        i++;
      }
    }

    return randomNumbers;
  };

  useEffect(() => {
    const intervalId = setInterval(() => {
      if (availableNumbers.length >= 3) {
        const duckNumbers = getRandomNumbers(availableNumbers, 3);
        setDuck1Number(duckNumbers[0]);
        setDuck2Number(duckNumbers[1]);
        setDuck3Number(duckNumbers[2]);

        setDuck1Position({
          x: Math.floor(Math.random() * (width-100)),
          y: Math.floor(Math.random() * (height-100)),
        });
        setDuck2Position({
          x: Math.floor(Math.random() * (width-100)),
          y: Math.floor(Math.random() * (height-100)),
        });
        setDuck3Position({
          x: Math.floor(Math.random() * (width-100)),
          y: Math.floor(Math.random() * (height-100)),
        });
      }
    }, 2000);

    return () => clearInterval(intervalId);
  }, [availableNumbers]);

  return (
    <div>
      <div
        className="moving-image"
        style={{ top: duck1Position.y, left: duck1Position.x }}
      >
        <Duck number={duck1Number} onClick={handleDuckClick} />
      </div>
      <div
        className="moving-image"
        style={{ top: duck2Position.y, left: duck2Position.x }}
      >
        <Duck number={duck2Number} onClick={handleDuckClick} />
      </div>
      <div
        className="moving-image"
        style={{ top: duck3Position.y, left: duck3Position.x }}
      >
        <Duck number={duck3Number} onClick={handleDuckClick} />
      </div>
    </div>
  );
};
