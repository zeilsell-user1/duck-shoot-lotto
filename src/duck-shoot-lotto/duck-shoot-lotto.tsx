import "./styles.css";

import { useEffect, useState } from "react";

import backgroundImage from "./../assets/duck-pod-back.jpg";
import foregroundImage from "./../assets/duck-pod-front.png";
import { Game } from "./game";

export const DuckShootLotto = (): JSX.Element => {
  const [gameStarted, setGameStarted] = useState(false);
  const [bgWidth, setBgWidth] = useState(0);
  const [bgHeight, setBgHeight] = useState(0);

  const handleStartGame = (): void => {
    setGameStarted(true);
  };

  const handleEndGame = (): void => {
    setGameStarted(false);
  };
  
  useEffect(() => {
    const backgroundImg = document.getElementById("background-img");
    setBgWidth(backgroundImg ? backgroundImg.clientWidth : 0);
    setBgHeight(backgroundImg ? backgroundImg.clientHeight : 0);
  }, [gameStarted]);

  return (
    <div className="board">
      <img id="background-img" className="background-image" src={backgroundImage} alt="Background image of pond" />
      <img className="top-image" src={foregroundImage} alt="Foreground image of pond vegitation" />

      {!gameStarted ? (
        <button className="controls" onClick={handleStartGame}>
          Start Game
        </button>
      ) : (
        <Game width={bgWidth} height={bgHeight} onGameEnd={handleEndGame} />
      )}
    </div>
  );
};
