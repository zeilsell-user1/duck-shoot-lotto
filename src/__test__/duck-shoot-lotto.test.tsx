import { DuckShootLotto } from "../duck-shoot-lotto";

describe("DuckShootLotto game component", () => {
  it("should render a default board", () => {
    const containerDefaultBoard = (): JSX.Element => <DuckShootLotto />;
    expect(containerDefaultBoard).toMatchSnapshot();
  });
});
