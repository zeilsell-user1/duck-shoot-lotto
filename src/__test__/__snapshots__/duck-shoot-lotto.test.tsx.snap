// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DuckShootLotto game component should render a default board 1`] = `[Function]`;
